### PR TITLE
Add PosixFilePath and friends support (for AFPP)

### PR DESCRIPTION
**Pull request by hasufell**
Depends on: https://github.com/haskell/filepath/pull/103

For testing simple file operations, you can use https://github.com/hasufell/file-io (also please review the [posix module](https://github.com/hasufell/file-io/blob/master/System/File/AbstractFilePath/Posix.hs))

@Bodigrim

### CODE REVIEW

File: cabal.project
Comment (hasufell) [inline cabal.project:11]: temporary

**Comment by Bodigrim** [comment]
@hasufell could you please rebase?

File: unix.cabal
Comment (Bodigrim) [inline unix.cabal:3]: Please keep version 2.8.0.0, the latest release was 2.7.

Comment (hasufell) [reply unix.cabal:3]: The idea was to bump beyond 2.8 on purpose, so that another major point release can be maintained in between that does not include AFPP.

Comment (Bodigrim) [reply unix.cabal:3]: Please stick to 2.8, we'll make amends if needed later.

**Comment by hasufell** [comment]
ping

File: cabal.project
Comment (hs-viktor) [inline cabal.project:3]: I am not entirely fond of manifest remote dependencies.  Ideally a build should be able to complete entire with locally staged inputs.  Is there some other way to signal the `file path` dependency?  Is this just a temporary measure?  When will this be resolved?

**Review by hs-viktor — COMMENTED** [review]
Looks fine with minor nits.  Please address those you don't object to, and rebase...

File: cabal.project
Comment (hasufell) [reply cabal.project:3]: This will be resolved once I actually publish filepath-2.x... that requires a few minor issues to be resolved: https://gitlab.haskell.org/haskell/filepath/-/issues?milestone_title=2.0.1

Comment (hs-viktor) [reply cabal.project:3]: Understood. I wouldn't want  to make a "release" with this unaddressed (unless unavoidable), but merging this as an interim commit, that could be used in a submodule, ... is fine if the consuming package deems it acceptable.  Does this work for you?

Any thoughts on the other comments?

Comment (hasufell) [reply cabal.project:3]: My plan is to make this PR mergable and then have it be merged after the filepath release, following a unix and Win32 release. The new filepath release cannot really be used without updated unix/Win32 releases.

Comment (hs-viktor) [reply cabal.project:3]: So, I take it the plan is to resolve all the issues, and then wait for the file path release?  You'll then ping the "unix" maintainers to push the button?

Comment (Bodigrim) [reply cabal.project:3]: Yes, precisely.

**Comment by Bodigrim** [comment]
@hasufell could you please rebase?

**Comment by vdukhovni** [comment]
Now that we have CI for "wasm", it is no longer in the way of routine work by wasm-agnostic maintainers. :-(
What's to be done about this?  Can the wasm CI be enabled only on wasm branches?  Or otherwise not get in the way?

**Comment by Bodigrim** [comment]
I’m AFK, but either Wasm CI job should honor settings in `cabal.project`, or `filepath` released officially. @vdukhovni are you happy with this branch otherwise?

**Comment by hs-viktor** [comment]
The ARM CI isn't working yet, and I should look a bit more carefully at the new Posix-flavoured Terminal.hsc.

**Comment by Bodigrim** [comment]
@TerrorJack `ci-wasm32-wasi` job does not seem to work any longer. Could you please take a look?

**Comment by TerrorJack** [comment]
Sure I'll open a PR against `master` to fix the job.

**Comment by hasufell** [comment]
Not due to a network failure: https://github.com/haskell/unix/runs/7280567595?check_suite_focus=true

I don't know what `wasm32-wasi-cabal`, but it clearly is not the same as `cabal`. Why do we have this even?

**Comment by TerrorJack** [comment]
It's now a cabal bound issue, since `ghc-wasm32-wasi` uses a `ghc-head` build. I'm looking into relaxing bounds in the wasm32 cabal.project file atm.

**Comment by hasufell** [comment]
We might have been hit by https://gitlab.haskell.org/ghc/ghc/-/issues/21738

**Comment by TerrorJack** [comment]
@hasufell Thanks for the pointer! Given `filepath` is vendored in `template-haskell` in latest head which should resolve this issue, I'll proceed to update `ghc-wasm32-wasi` to latest head to unblock CI, but that'll take a couple more hours, are you fine with this timeframe?

**Comment by TerrorJack** [comment]
To provide more context, the job is to added to test `unix` support for `wasm32-wasi` platform, upcoming in 9.6, with more details in https://gitlab.haskell.org/ghc/ghc/-/wikis/WebAssembly-backend

**Comment by TerrorJack** [comment]
If you push the fix at https://gist.github.com/TerrorJack/6352f35dc53e7b1b1baef4391ff86c3d, the `ci-wasm32-wasi` job should have been fixed now.

**Review by hs-viktor — APPROVED** [review]
LGTM. Modulo a pending rebase and ideally a successful CI run.

File: System/Posix/IO/PosixString.hsc
Comment (hasufell) [inline System/Posix/IO/PosixString.hsc:67]: We import `fdRead` and `fdWrite` from ByteString now.

**Comment by Bodigrim** [comment]
@hasufell there are several redundant imports. GHC build will choke on warnings.

**Review by Bodigrim — APPROVED** [review]
Thanks, great!

**Pull request merged**
